### PR TITLE
API-1673: Add description annotations to service CA secret and configmap

### PR DESCRIPTION
**Pull request by dinhxuanvu**
Service CA generates a secret and configmap for its signing key.
These secret and configmap need to have description annotation
to conform to the new cert ownership validation.

For all secrets that are created by service CA for services, an
ownership and description annotations are generated to indicate
service CA owns these secrets and its corresponding service owner.

Signed-off-by: Vu Dinh <vudinh@outlook.com>

### CODE REVIEW

**Comment by dinhxuanvu** [comment]
@deads2k @vrutkovs PTAL

File: pkg/controller/servingcert/controller/secret_creating_controller.go [hunk 1]
```diff
@@ -433,3 +431,25 @@ func uidsEqual(secret *corev1.Secret, service *corev1.Service) bool {
...
+// Set ownership and description annotations. Return true if there are changes on
+// the two annotations
+func setSecretOwnerDescription(secret *corev1.Secret, service *corev1.Service) bool {
+	changed := false
+	if secret.Annotations == nil {
+		secret.Annotations = map[string]string{}
+		changed = true
+	}
+	// Set service CA owner annotation to generated secret
+	if len(secret.Annotations[apiannotations.OpenShiftDescription]) == 0 {
```
Comment (vrutkovs) [inline pkg/controller/servingcert/controller/secret_creating_controller.go:444]:
```suggestion
	if len(secret.Annotations[apiannotations.OpenShiftComponent]) == 0 {
```

File: bindata/v4.0.0/controller/signing-cabundle.yaml [hunk 2]
```diff
@@ -5,5 +5,6 @@ metadata:
   name: signing-cabundle
   annotations:
     openshift.io/owning-component: service-ca
+    openshift.io/description: "Service CA signing CA bundle"
```
Comment (deads2k) [inline bindata/v4.0.0/controller/signing-cabundle.yaml:8]: This should be more descriptive. What is it used by, for what purpose.

File: pkg/controller/servingcert/controller/secret_creating_controller.go [hunk 3]
```diff
@@ -433,3 +431,25 @@ func uidsEqual(secret *corev1.Secret, service *corev1.Service) bool {
...
+	}
+	// Set service CA owner annotation to generated secret
+	if len(secret.Annotations[apiannotations.OpenShiftComponent]) == 0 {
+		secret.Annotations[apiannotations.OpenShiftComponent] = api.OwningJiraComponent
+		changed = true
+	}
+	// Copy description annotation from service to secret if exists. Otherwise, generate
+	// a description for it <service name>/<service namespace>
+	if len(secret.Annotations[apiannotations.OpenShiftDescription]) == 0 {
+		secret.Annotations[apiannotations.OpenShiftDescription] = fmt.Sprintf("Generated by Service CA for %s/%s", service.Name, service.Namespace)
```
Comment (deads2k) [inline pkg/controller/servingcert/controller/secret_creating_controller.go:451]: indicate what you'd use this for and how someone could verify it.

**Comment by deads2k** [comment]
configmaps with injected CA bundles need descriptions (if not present) too.  They should describe what sorts of certificates they can be used to verify.

File: bindata/v4.0.0/controller/signing-cabundle.yaml (hunk 2 above)
Comment (dinhxuanvu) [reply bindata/v4.0.0/controller/signing-cabundle.yaml]: Fixed.

File: bindata/v4.0.0/controller/signing-secret.yaml [hunk 4]
```diff
@@ -5,6 +5,7 @@ metadata:
   name: signing-key
   annotations:
     openshift.io/owning-component: service-ca
+    openshift.io/description: "Service CA signing key"
```
[the remark this answers is not in the document]
Comment (dinhxuanvu) [reply bindata/v4.0.0/controller/signing-secret.yaml]: Fixed

**Comment by dinhxuanvu** [comment]
@deads2k PTAL. Thanks

File: pkg/controller/servingcert/controller/secret_creating_controller.go [hunk 5]
```diff
@@ -433,3 +431,24 @@ func uidsEqual(secret *corev1.Secret, service *corev1.Service) bool {
...
+		changed = true
+	}
+	// Set service CA owner annotation to generated secret
+	if len(secret.Annotations[apiannotations.OpenShiftComponent]) == 0 {
+		secret.Annotations[apiannotations.OpenShiftComponent] = api.OwningJiraComponent
+		changed = true
+	}
+	// Generate a description for generated secret if not existed already
+	if len(secret.Annotations[apiannotations.OpenShiftDescription]) == 0 {
+		secret.Annotations[apiannotations.OpenShiftDescription] = fmt.Sprintf("Secret contains a pair signed serving certificate/key that is generated by Service CA operator for service %s/%s and is annotated to the service with annotating a service resource with 'service.beta.openshift.io/serving-cert-secret-name: %s'. The certificate is valid for 2 years.", service.Name, service.Namespace, secret.Name)
```
Comment (deads2k) [inline pkg/controller/servingcert/controller/secret_creating_controller.go:450]:
```suggestion
		secret.Annotations[apiannotations.OpenShiftDescription] = fmt.Sprintf("Secret contains a pair signed serving certificate/key that is generated by Service CA operator for service/%s with hostname %s.%s.svc and is annotated to the service with annotating a service resource with 'service.beta.openshift.io/serving-cert-secret-name: %s'. The certificate is valid for 2 years.", service.Name,  service.Name, service.Namespace, secret.Name)
```

willing to take as followup.

File: pkg/controller/servingcert/controller/secret_updating_controller.go [hunk 6]
```diff
@@ -210,8 +209,7 @@ func (sc *serviceServingCertUpdateController) ensureSecretData(service *v1.Servi
 	}
 
 	// set the owning-component unless someone else has claimed it.
-	if !update && len(secretCopy.Annotations[apiannotations.OpenShiftComponent]) == 0 {
-		secretCopy.Annotations[apiannotations.OpenShiftComponent] = api.OwningJiraComponent
+	if !update && setSecretOwnerDescription(secretCopy, service) {
```
Comment (deads2k) [inline pkg/controller/servingcert/controller/secret_updating_controller.go:212]:
```suggestion
	needsOwnerUpdate := setSecretOwnerDescription(secretCopy, service) 
	if !update && needsOwnerUpdate {
```

slightly easier for reviewer to read.

Comment (dinhxuanvu) [reply pkg/controller/servingcert/controller/secret_updating_controller.go]: Done

**Pull request merged**